API.md

```diff
@@ -0,0 +1,93 @@
+# Introduction
+This document provides API specification of NodeResourceTopology API which is a CRD
+based API that describes the structure of topology on a node. One of the goals when
+designing this API was to ensure hardware independent data structure, which allows
+us to have a future-proof API that represents hardware of varying internal topology.
+
+CRD instances must be created external to kubelet and independently. Each instance
+will be bound to one working node.
+
+# NodeTopology Object Definition
+
+```
+// NodeResourceTopology describes node resources and their topology.
+type NodeResourceTopology struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Zones      ZoneList      `json:"zones"`
+	Attributes AttributeList `json:"attributes,omitempty"`
+}
+
+// Zone represents a resource topology zone, e.g. socket, node, die or core.
+type Zone struct {
+	Name       string           `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type       string           `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Parent     string           `json:"parent,omitempty" protobuf:"bytes,3,opt,name=parent"`
+	Costs      CostList         `json:"costs,omitempty" protobuf:"bytes,4,rep,name=costs"`
+	Attributes AttributeList    `json:"attributes,omitempty" protobuf:"bytes,5,rep,name=attributes"`
```

Comment:
If there are more than one device on a NUMA node and these device can be different for some attributes, how should we set attributes for these devices?

Reply:
These attributes are for the zone and cannot represent attributes that differ per-resource. For that, we'd need to extend the `ResourceInfo` type, I think.

@swatisehgal @ffromani

Reply:
For resource level attributes, extending ResourceInfo is certainly one way. This seems like a reasonable addition to me.

Another way could be to create a new zone that is child of a zone representing a NUMA node and then capture these resources  with their corresponding attributes under the newly created sub-NUMA zone.

The former requires some changes while the latter doesn't. I don't feel too strongly about either so would like to hear @ffromani and @marquiz opinion on this.

Reply:
> Another way could be to create a new zone that is child of a zone representing a NUMA node and then capture these resources with their corresponding attributes under the newly created sub-NUMA zone.

True, this should work.
 
> The former requires some changes while the latter doesn't. I don't feel too strongly about either so would like to hear @ffromani and @marquiz opinion on this.

I don't have strong opinions either. Resorting to the latter (no changes) until we understand the usage scenarios for the former(?)